File: adapters/python/lexicon_python/emission.py
"""Canonical Lexicon record ordering and JSONL emission."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

from .contract import LANGUAGE, SCHEMA_VERSION
from .model import Facts


def _span_key(record: dict[str, Any]) -> tuple[Any, ...]:
    value = record.get("span") or {}
    return (
        value.get("path", ""),
        value.get("start_line", 0),
        value.get("start_column", 0),
        value.get("end_line", 0),
        value.get("end_column", 0),
    )


def _record_sort_key(record: dict[str, Any]) -> tuple[Any, ...]:
    kind = record["record"]
    if kind == "node":
        return (0, record["id"], record["kind"], record["path"], record["qualified_name"])
    if kind == "edge":
        return (1, record["source"], record["target"], record["relation"], *_span_key(record))
    return (
        2,
        record["source"],
        record["relation"],
        record["expression"],
        record["reason"],
        *_span_key(record),
    )


def emit_records(facts: Facts, adapter_version: str) -> list[dict[str, Any]]:
    header = {
        "record": "lexicon",
        "schema_version": SCHEMA_VERSION,
        "adapter_version": adapter_version,
        "language": LANGUAGE,
        "repository": facts.repository,
    }
    records = [header, *sorted(facts.nodes.values(), key=_record_sort_key)]
    records.extend(sorted(facts.edges.values(), key=_record_sort_key))
    records.extend(sorted(facts.unresolved.values(), key=_record_sort_key))
    return records


def write_records(records: list[dict[str, Any]], output: Path) -> None:
    lines = [json.dumps(record, ensure_ascii=False, sort_keys=True, separators=(",", ":")) for record in records]
    if str(output) == "-":
        sys.stdout.write("\n".join(lines) + "\n")
        return
    destination = output.expanduser()
    destination.parent.mkdir(parents=True, exist_ok=True)
    destination.write_text("\n".join(lines) + "\n", encoding="utf-8", newline="\n")

File: adapters/typescript/src/emission.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { compareKeys, factSortKey, jsonLine, LANGUAGE, SCHEMA_VERSION, VERSION } from "./contract";
import type { Fact, FactStore } from "./model";

export function emitFacts(facts: FactStore): Fact[] {
  const header: Fact = {
    adapter_version: VERSION,
    language: LANGUAGE,
    record: "lexicon",
    repository: facts.repository,
    schema_version: SCHEMA_VERSION,
  };
  const nodes = Array.from(facts.nodes.values()).sort((a, b) => compareKeys(factSortKey(a), factSortKey(b)));
  const edges = Array.from(facts.edges.values()).sort((a, b) => compareKeys(factSortKey(a), factSortKey(b)));
  const unresolved = Array.from(facts.unresolved.values()).sort((a, b) => compareKeys(factSortKey(a), factSortKey(b)));
  return [header, ...nodes, ...edges, ...unresolved];
}

export function writeJsonl(records: Fact[], outputPath: string): void {
  const lines = records.map(jsonLine).join("\n") + "\n";
  if (outputPath === "-") {
    process.stdout.write(lines);
    return;
  }
  const destination = path.resolve(outputPath);
  fs.mkdirSync(path.dirname(destination), { recursive: true });
  fs.writeFileSync(destination, lines, { encoding: "utf8" });
}
